### PR TITLE
[MAJOR] EDA-1360 logs view

### PR DESCRIPTION
**Pull request by ibrizuela-eb**
-View to visualize match details in an understandable way to humans.

### CODE REVIEW

File: development/server_requests.py [hunk 1]
```diff
@@ -15,3 +14,14 @@ def send_challenge(*args, **kwargs):
         '{}:{}/challenge'.format(SERVER_URL, SERVER_PORT),
         json=data,
     )
+
+
+def get_logs(*args, **kwargs):
```
Comment (nicolasnunez-eb) [inline development/server_requests.py:19]: Why not passing explicit arguments like `get_logs(game_id, page_token, requests)` ?
If someone wants to use this function he will not know what this function expects

File: development/tests/test_server_requests.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    class MockResponse:
+        def __init__(self, json_data, status_code):
+            self.json_data = json_data
+            self.status_code = status_code
+
+        def json(self):
+            return self.json_data
+
+    return MockResponse(data, 200)
```
Comment (nicolasnunez-eb) [inline development/tests/test_server_requests.py:22]: I think it is not necessary to do all of this there is a more straightforward way. We can talk about that in slack

File: development/views_api.py [hunk 3]
```diff
@@ -39,7 +39,7 @@ def convert_data(req_data):
     """ Recieve a dictionary and return a new dictionary. """
```
Comment (nicolasnunez-eb) [inline development/views_api.py:39]: This is very vague, I know it is not part of this PR but you can specify a little more about what is going on here

File: development/__init__.py [hunk 4]
```diff
@@ -1 +1,5 @@
-ACTIONS = '''<a href="{% url 'development:match_detail' record.pk %}">view</a>'''
+from environment import get_env_variable
+
+
+SERVER_URL = get_env_variable('SERVER_URL')
+SERVER_PORT = get_env_variable('SERVER_PORT')
```
Comment (eb-cmoyano) [inline development/__init__.py:5]: Could you move these vars to Django settings?

https://github.com/evbeda/edagames-django/blob/main/edagames/settings.py

File: development/server_requests.py [hunk 5]
```diff
@@ -1,8 +1,7 @@
-from environment import get_env_variable
-
-
-SERVER_URL = get_env_variable('SERVER_URL')
-SERVER_PORT = get_env_variable('SERVER_PORT')
+from development import (
+    SERVER_PORT,
+    SERVER_URL,
+)
```
Comment (eb-cmoyano) [inline development/server_requests.py:4]: Later:
```
from django.conf import settings
```

File: development/server_requests.py [hunk 6]
```diff
@@ -15,3 +14,14 @@ def send_challenge(*args, **kwargs):
...
     )
+
+
+def get_logs(*args, **kwargs):
+    data = {
+        "game_id": kwargs['game_id'],
+        "page_token": kwargs['page_token'],
+    }
+    return kwargs['requests'].get(
+        '{}:{}/match_details'.format(SERVER_URL, SERVER_PORT),
```
Comment (eb-cmoyano) [inline development/server_requests.py:25]: nit: here you could use f strings.

File: development/templates/development/match_details.html [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            <tr>
+                <td>{{ log.state }}</td>
+                <td>{{ log.current_player }}</td>
+                <td>{{ log.player_1 }}</td>
+                <td>{{ log.score_1 }}</td>
+                <td>{{ log.player_2 }}</td>
+                <td>{{ log.score_2 }}</td>
+                <td>({{ log.from_row }}, {{ log.from_col }})</td>
+                <td>({{ log.to_row }}, {{ log.to_col }})</td>
+                <td style="font-family:'sans-serif', monospace">{{ log.board }}</td>
```
Comment (eb-cmoyano) [inline development/templates/development/match_details.html:36]: nit: ideally, we should apply styles in a CSS class.

File: development/server_requests.py [hunk 8]
```diff
@@ -15,3 +14,14 @@ def send_challenge(*args, **kwargs):
         '{}:{}/challenge'.format(SERVER_URL, SERVER_PORT),
```
Comment (eb-cmoyano) [inline development/server_requests.py:14]: What????

File: development/tests/test_server_requests.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+
+class TestServerRequests(TestCase):
+
+    @patch(
+        'requests.get',
+        side_effect=lambda x: mocked_requests_get(
```
Comment (eb-cmoyano) [inline development/tests/test_server_requests.py:29]:
```
import json
from unittest.mock import MagicMock

mocked_response = MagicMock(json=lambda x: json.loads(x), spec=["json"])

with patch('requests.get', return_value=mocked_respons) as request_get_mocked:
    ....
    request_get_mocked.assert_called_once_with(...)
```
I think you could simply use `return_value=mocked_response`

File: development/tests/test_server_requests.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        ),
+    )
+    def test_send_challenge(self):
```
Comment (eb-cmoyano) [inline development/tests/test_server_requests.py:32]: Please, also add a test for `get_logs` as well.

File: development/views.py [hunk 11]
```diff
@@ -52,18 +56,62 @@ def form_valid(self, form):
...
     model = Match
-    table_class = MatchTable
     template_name = 'development/match_history.html'
 
     def get_queryset(self):
-        return Match.objects.filter(user_1=self.request.user) | Match.objects.filter(user_2=self.request.user)
+        return Match.objects.filter(
+            Q(user_1=self.request.user)
+            | Q(user_2=self.request.user)
+        ).order_by("-date_match")
```
Comment (eb-cmoyano) [inline development/views.py:67]: Performance: Please use `.union` instead.

Reference:
https://docs.djangoproject.com/en/2.2/ref/models/querysets/#when-querysets-are-evaluated

File: development/views.py [hunk 12]
```diff
@@ -52,18 +56,62 @@ def form_valid(self, form):
...
+    def __init__(self, *args, **kwargs):
+        super(MatchDetailView, self).__init__(*args, **kwargs)
+        self.current_page = 1
+        self.prev_page = 1
+        self.next_page = 2
+        self.pages = {
+            1: '',
+            2: '',
+            3: '',
+        }
```
Comment (eb-cmoyano) [inline development/views.py:82]: If this a work in progress please add a `# TODO: Think about how to deal with pagination ` comment.

File: development/views.py [hunk 13]
```diff
@@ -52,18 +56,62 @@ def form_valid(self, form):
...
+            self.current_page = page
+            self.prev_page = (
+                page - 1
+                if (page - 1) > 1
+                else 1
+            )
+            self.next_page = page + 1
+
+        response = get_logs(
+            requests=requests,
```
Comment (eb-cmoyano) [inline development/views.py:99]: nooo! 😢

File: development/server_requests.py (hunk 5 above)
Comment (ibrizuela-eb) [reply development/server_requests.py]: done!

File: development/__init__.py (hunk 4 above)
Comment (ibrizuela-eb) [reply development/__init__.py]: done!

File: development/server_requests.py (hunk 1 above)
Comment (ibrizuela-eb) [reply development/server_requests.py]: done!

File: development/server_requests.py (hunk 6 above)
Comment (ibrizuela-eb) [reply development/server_requests.py]: done!

File: development/templates/development/match_details.html (hunk 7 above)
Comment (ibrizuela-eb) [reply development/templates/development/match_details.html:36]: done!

File: development/tests/test_server_requests.py (hunk 9 above)
Comment (ibrizuela-eb) [reply development/tests/test_server_requests.py]: done!

File: development/server_requests.py (hunk 8 above)
Comment (ibrizuela-eb) [reply development/server_requests.py]: I'm sorry you've had to see that.

File: development/tests/test_server_requests.py (hunk 10 above)
Comment (ibrizuela-eb) [reply development/tests/test_server_requests.py:13]: done!

File: development/tests/test_server_requests.py (hunk 2 above)
Comment (ibrizuela-eb) [reply development/tests/test_server_requests.py]: It was removed.

File: development/views_api.py (hunk 3 above)
Comment (ibrizuela-eb) [reply development/views_api.py]: done!

File: development/views.py (hunk 11 above)
Comment (ibrizuela-eb) [reply development/views.py]: done!

File: development/views.py (hunk 12 above)
Comment (ibrizuela-eb) [reply development/views.py]: done!

File: development/views.py (hunk 13 above)
Comment (ibrizuela-eb) [reply development/views.py]: It was removed.

**Review by vgatica-eb — APPROVED** [review]

**Pull request merged**